### PR TITLE
Make Evaluation.State an IntEnum

### PR DESCRIPTION
**Pull request by fekoch**
Originally a part of PR #2136. Like @niklasmohrin suggested, I pulled the specific changes required to make `Evaluation.State` an enum.

In the process I discovered Django's own `models.IntegerChoices`, which works even better than the python `IntEnum`. It also generates a migration that does not depend on the `Evaluation` class (also an issue @niklasmohrin mentioned)

### CODE REVIEW

File: evap/evaluation/tools.py
Comment (niklasmohrin) [inline evap/evaluation/tools.py:329]: This is not used anywhere anymore :)

File: evap/staff/views.py
Comment (niklasmohrin) [inline evap/staff/views.py:497]: We can also mark the return type as `Evaluation.State` instead of `int` then :)

File: evap/staff/tests/test_views.py
Comment (niklasmohrin) [inline evap/staff/tests/test_views.py:2233]: The new test does not contain any `assertNotIn`, so it is very well possible that "prepared", "new", etc. come up accidentally. Do you think you can come up with a way to avoid this, or is it too annoying now that Django handles the translation?

File: evap/staff/tests/test_views.py
Comment (niklasmohrin) [inline evap/staff/tests/test_views.py:2244]: I think we can lean a bit more towards integration-testing and use "neu" etc. directly

File: evap/evaluation/tests/test_views.py
Comment (niklasmohrin) [inline evap/evaluation/tests/test_views.py:285]: CI complains about "TODO" - let's remove these for now and keep this PR unrelated to your other one :)

File: evap/evaluation/models.py
Comment (niklasmohrin) [inline evap/evaluation/models.py:1003]: Nice, we can get rid of `transform_log_action` everywhere then! If you want, you can try it here, otherwise you or someone else can also do it in another PR later

File: evap/evaluation/models.py
Comment (niklasmohrin) [inline evap/evaluation/models.py:787]: I suppose we don't need this anymore, the one usage can just do `evaluation.state.label` itself

File: evap/evaluation/tests/test_views.py
Comment (niklasmohrin) [inline evap/evaluation/tests/test_views.py:273]: Maybe you can refactor this to take `success_expected: bool` instead of `assertion` and do the `assertEqual` etc. here?

**Review by niklasmohrin — COMMENTED** [review]
Very cool, some comments:

File: evap/staff/tests/test_views.py
Comment (niklasmohrin) [reply evap/staff/tests/test_views.py:2233]: (I think the better translation is worth more than the test here, so I am not expecting any changes outside of tests for this comment)

File: evap/staff/tests/test_views.py
Comment (richardebeling) [reply evap/staff/tests/test_views.py:2244]: I'm not sure if we've ever done that (make tests rely on the translation stored in messages.po). For some other tests, I think we mock `gettext` or other translation functions to make them return some unique string. The tests then
* look for the existence of the unique string on the page
* assert that the mock was called with the right input

This might help with https://github.com/e-valuation/EvaP/pull/2174#discussion_r1585330486 above, but it might also be overkill. If using the translations stored in messages.mo helps to keep it simple here, that's also fine with me

File: evap/evaluation/models.py
Comment (fekoch) [reply evap/evaluation/models.py:787]: I believe we can't. The field `state` can be just an integer so we need to convert it into an enum before calling .label.

Comment (fekoch) [reply evap/evaluation/models.py:787]: We could of course call `Evaluation.State(evaluation.state).label` but I believe this is verbose enough to warrant it's own function.

But if you have a good way to make sure that `state` is always set to an enum member, this would be even better

File: evap/evaluation/models.py
Comment (fekoch) [reply evap/evaluation/models.py:1003]: we should probably move this into a different PR. When I quickly tested it just now, it threw multiple issues.

File: evap/staff/tests/test_views.py
Comment (fekoch) [reply evap/staff/tests/test_views.py:2256]: nevermind. It seems like this mocking persists past the end of the test.

Comment (fekoch) [reply evap/staff/tests/test_views.py]: I could not get translation with mocking to work and decided to just match for the full html fragment instead (like you @niklasmohrin thought, the response of course contained the string "new" :) )

File: evap/evaluation/models.py
Comment (niklasmohrin) [reply evap/evaluation/models.py:787]: We could rename the database field to `_state` and add

```python
@property
def state(self):
    return Evaluation.State(self._state)
```

Not sure if we would run into some problems though. @richardebeling What do you think? It sounds like a fair change to me

File: evap/evaluation/tests/test_views.py
Comment (niklasmohrin) [inline evap/evaluation/tests/test_views.py:274]: What do you think about the following?

```suggestion
        confirmation_form.submit(expect_errors=not success_expected)

        if success_expected:
            self.assertEqual(Evaluation.objects.get(pk=evaluation.pk).state, Evaluation.State.NEW)
```

Comment (niklasmohrin) [reply evap/evaluation/tests/test_views.py:274]: This also leaves the question whether we want to do anything in the `else` case - I think it would also be fine like this

File: evap/evaluation/models.py
Comment (Kakadus) [reply evap/evaluation/models.py:787]: I also used this pattern and it worked quite well :+1:

Comment (richardebeling) [reply evap/evaluation/models.py:788]: As pointed out by @fekoch today: The property has the disadvantage that using the field in django db field specifications (`evaluation__state`) becomes weird (triple underscore? Does that work with the parsing?) when using django's ORM or model bakery.

I'd be fine with all proposed solutions as long as you can still use the field with django DB field lookups.

File: evap/evaluation/tests/test_views.py
Comment (richardebeling) [inline evap/evaluation/tests/test_views.py:256]: A bit annoying, but pep-484 defaults to "Any" instead of "None" for methods that return nothing
```suggestion
    def setUpTestData(cls) -> None:
```

File: evap/evaluation/tests/test_views.py
Comment (richardebeling) [inline evap/evaluation/tests/test_views.py:260]:
```suggestion
    def reset_from_x_to_new(self, x, success_expected: bool) -> None:
```

File: evap/staff/tests/test_views.py
Comment (richardebeling) [inline evap/staff/tests/test_views.py:2243]: I've observed `html=True` to be annoyingly slow in the past -- do we need it here? What's the average test runtime with and without the flag?

File: evap/evaluation/tests/test_views.py
Comment (richardebeling) [reply evap/evaluation/tests/test_views.py:274]: (Agreed. `"Should have raised exception by now"` seems subtly wrong to me, exceptions wouldn't propagate into the test context here. I would have expected this to fail -- maybe the `assertRaises` below is even triggered by some unintended error)

File: evap/staff/tests/test_views.py
Comment (fekoch) [reply evap/staff/tests/test_views.py:2243]: comparing:
| Header | Header |
|--------|--------|
| `self.assertContains(response, "<li> TRANSLATED-state: TRANSLATED-new &#8594; TRANSLATED-prepared </li>", html=True)` | ~0.140s |
| `self.assertContains(response, "TRANSLATED-state: TRANSLATED-new &#8594; TRANSLATED-prepared", html=True)` | ~0.154s |
| `self.assertContains(response, "TRANSLATED-state: TRANSLATED-new &#8594; TRANSLATED-prepared")` | ~0.132s |

Yeah, I think without `html=True` it works just as well

File: evap/evaluation/models.py
Comment (fekoch) [reply evap/evaluation/models.py:784]: I would argue for just keeping the `state` field as is. I decided for keeping the `state_str()` property which already existed for the one instance where it is used.

File: evap/evaluation/tests/test_views.py
Comment (fekoch) [reply evap/evaluation/tests/test_views.py]: I changed it to:
```python
        confirmation_page = form.submit(name="target_state", value=str(Evaluation.State.NEW))

        if success_expected:
            confirmation_form = confirmation_page.forms["evaluation-operation-form"]
            confirmation_form.submit()
            self.assertEqual(Evaluation.objects.get(pk=evaluation.pk).state, Evaluation.State.NEW)
        else:  # no confirmation form should exist
            self.assertNotIn("evaluation-operation-form", confirmation_page.forms)

```

File: evap/evaluation/templatetags/evaluation_filters.py
Comment (fekoch) [inline evap/evaluation/templatetags/evaluation_filters.py:25]: I could not find any usages of this dict and removing it did not break any tests, is there a reason to keep it around?

Comment (richardebeling) [reply evap/evaluation/templatetags/evaluation_filters.py:25]: I don't think so, probably left unused after some past change, removing should be fine. Git's pickaxe functions (`git log -S STR_TO_STATE`) might show you whether/when the last usage was removed.

File: evap/evaluation/models.py
Comment (niklasmohrin) [reply evap/evaluation/models.py:784]: Ah, the ORM story is a good point. Okay then, let's keep it as is :)

File: evap/evaluation/tests/test_views.py
Comment (niklasmohrin) [inline evap/evaluation/tests/test_views.py:278]: (Nit) I think we can be more succinct here:
```suggestion
        if success_expected:
            confirmation_page.forms["evaluation-operation-form"].submit()
            self.assertEqual(Evaluation.objects.get(pk=evaluation.pk).state, Evaluation.State.NEW)
        else:
            self.assertNotIn("evaluation-operation-form", confirmation_page.forms)
```

File: evap/staff/tests/test_views.py
Comment (niklasmohrin) [inline evap/staff/tests/test_views.py:2235]: Can we keep the `assertNotIn` check before the call to `ready_for_editors`?

File: evap/evaluation/models.py
Comment (niklasmohrin) [inline evap/evaluation/models.py:387]: (Maybe follow up work)

@richardebeling @Kakadus 

The string `editor_approved` is still not intended for display on the website (we have the `statename` template filter for that) - maybe we can get rid of it at some point and just encode it here? We have other `*Choices` where we use whitespace in the string id (for example `GradeDocument.Type`). I suppose we should come up with a getter that always returns `Evaluation.State` then (not like our current `evaluation.state` which might return `int`)

Is there any reason to not replace `_` with ` ` already now? If not, we could do it now to possibly avoid an additional migration later

**Review by niklasmohrin — APPROVED** [review]
This turned out really nicely :)

I haven't checked whether the migration is up to date and leave that to the final reviewer :^)

File: evap/evaluation/models.py
Comment (richardebeling) [reply evap/evaluation/models.py:387]: > Is there any reason to not replace _ with  already now?

None that I know of, putting in the user-facing string is fine with me 👍

File: evap/evaluation/templatetags/evaluation_filters.py
Comment (fekoch) [reply evap/evaluation/templatetags/evaluation_filters.py:25]: @niklasmohrin you added this in 2021, but even at that commit (https://github.com/e-valuation/EvaP/tree/37508eb2c426d1e61437e24832a12dbad0d95d12) I do not find any usages

Comment (niklasmohrin) [reply evap/evaluation/templatetags/evaluation_filters.py:25]: :thinking: :thinking: :thinking: 
image

Huh, yeah, I guess it's fair to delete it then :D

**Review by Kakadus — APPROVED** [review]

File: evap/evaluation/models.py
Comment (Kakadus) [reply evap/evaluation/models.py:387]: searching for `_\("[\w_ ]*_[\w_ ]*"\)` does not yield any other translations, so I'm fine with fixing this here

Comment (richardebeling) [reply evap/evaluation/models.py:387]: Note: The same is probably true for `in_evaluation`. Not having another migration doesn't seem so important to me that this PR needs to be blocked by that. I'd say this can be part of another PR.

File: evap/evaluation/tests/test_views.py
Comment (richardebeling) [inline evap/evaluation/tests/test_views.py:271]: (opinion): If you add `cls.url = reverse("staff:semester_view", args=[self.semester.pk])` in `setUpTestData`, this can be compacted to:

```suggestion
    def reset_from_x_to_new(self, x, success_expected: bool) -> None:
        evaluation = baker.make(Evaluation, state=x, course__semester=self.semester)

        semester_overview_page = self.app.get(self.url, user=self.manager, status=200 )
        form = semester_overview_page.forms["evaluation_operation_form"]
        form["evaluation"] = [evaluation.pk]
        confirmation_page = form.submit(name="target_state", value=str(Evaluation.State.NEW))
```

which I like more -- imo multiple single-statement "blocks" don't really help with readability as they don't provide semantical grouping.

File: evap/evaluation/tests/test_views.py
Comment (richardebeling) [inline evap/evaluation/tests/test_views.py:277]: To me, it would be natural to assert that the evaluation state hasn't changed in the else case -- is there a reason you didn't do that?

At the same time, the fact that the form doesn't exist anymore seems more like a accidental implementation detail (I guess we redirect somewhere?) that might change.

File: evap/evaluation/tests/test_views.py
Comment (richardebeling) [inline evap/evaluation/tests/test_views.py:258]: is `results_are_archived` relevant? If so, maybe add a short comment why? I find it a bit confusing because on a real instance, you wouldn't want to change evaluation states of archived evaluations / in archived semesters

File: evap/evaluation/tests/test_views.py
Comment (richardebeling) [inline evap/evaluation/tests/test_views.py:291]: I'd add the parameter names here to make reading easier:

```suggestion
        for s in valid_start_states:
            self.reset_from_x_to_new(s, success_expected=True)
        for s in invalid_start_states:
            self.reset_from_x_to_new(s, success_expected=False)
```

File: evap/evaluation/tests/test_views.py
Comment (fekoch) [reply evap/evaluation/tests/test_views.py:273]: I replaced the check for the nonexisting form with a comparison of the HTTP Status code and assert that must not be 200. Or is it better to not do that check?

File: evap/evaluation/tests/test_views.py
Comment (fekoch) [reply evap/evaluation/tests/test_views.py]: I do not think this is relevant. The original commit was probably in January or something

**Pull request merged**
